File: frontend/src/lib/post/adapter/gateway/updatePostRequestDto.ts
import type { UpdatePostParams } from '$lib/post/application/gateway/postRepository';

export class UpdatePostRequestDto {
	title: string;
	description: string;
	content: string;
	labelIds: number[];
	previewImageUrl: string | null;
	publishedTime: string | null;

	private constructor(props: {
		title: string;
		description: string;
		content: string;
		labelIds: number[];
		previewImageUrl: string | null;
		publishedTime: string | null;
	}) {
		this.title = props.title;
		this.description = props.description;
		this.content = props.content;
		this.labelIds = props.labelIds;
		this.previewImageUrl = props.previewImageUrl;
		this.publishedTime = props.publishedTime;
	}

	static fromParams(params: UpdatePostParams): UpdatePostRequestDto {
		return new UpdatePostRequestDto({
			...params,
			previewImageUrl: params.previewImageUrl?.href ?? null,
			publishedTime: params.publishedTime?.toISOString() ?? null,
		});
	}

	toJson() {
		return {
			title: this.title,
			description: this.description,
			content: this.content,
			label_ids: this.labelIds,
			preview_image_url: this.previewImageUrl,
			published_time: this.publishedTime,
		};
	}
}

File: backend/feature/post/src/adapter/delivery/update_post_request_dto.rs
use chrono::{DateTime, Utc};
use serde::Deserialize;
use utoipa::ToSchema;

use crate::{
    application::gateway::update_post_params::UpdatePostParams,
    application::service::image_extractor::ImageExtractor,
};

#[derive(Deserialize, ToSchema, Clone)]
pub struct UpdatePostRequestDto {
    pub title: String,
    pub description: String,
    pub content: String,
    pub label_ids: Vec<i32>,

    #[schema(required, format = Uri)]
    pub preview_image_url: Option<String>,

    #[schema(required, format = DateTime)]
    pub published_time: Option<String>,
}

impl From<UpdatePostRequestDto> for UpdatePostParams {
    fn from(val: UpdatePostRequestDto) -> Self {
        let image_ids = ImageExtractor::extract_image_ids(&val.content, &val.preview_image_url);

        UpdatePostParams {
            title: val.title,
            description: val.description,
            content: val.content,
            label_ids: val.label_ids,
            preview_image_url: val.preview_image_url,
            published_time: val
                .published_time
                .and_then(|time_str| DateTime::parse_from_rfc3339(&time_str).ok())
                .map(|dt| dt.with_timezone(&Utc)),
            image_ids,
        }
    }
}
